### ISSUE TITLE
[FEATURE] StarCraft Multi-Agent Challenge v2 (SMACv2) Environment Support

### ISSUE THREAD

**Issue by Amr92Mamdouh** [Abdulhamid97Mousa/mosaic#27]
MOSAIC supports a wide range of multi-agent environments but does not currently include SMACv2 (Ellis et al., NeurIPS 2023), the successor to SMAC and the most rigorous cooperative MARL benchmark available today. SMACv2 was designed specifically to fix the evaluation blind spots in SMAC v1 —where agents could learn near-optimal policies by memorising fixed map layouts rather than by generalising. Integrating SMACv2 would give MOSAIC researchers a benchmark that actually measures cooperative generalisation, not map memorisation.

###  What is SMACv2?
SMACv2 is a cooperative multi-agent benchmark built on top of StarCraft II via the pysc2 library. It extends SMAC v1 with procedural generation:
- Unit start positions are randomised every episode (agents cannot memorise formation strategies)
- Unit type compositions are randomised across a distribution (agents face novel team configurations each episode)
- 3 procedural maps: protoss, terran, zerg — each with configurable team sizes (5v5, 10v10, 20v20)
- Same core interface as SMAC: per-agent partial observations, discrete action space (no-op, move, attack, stop), global state tensor for CTDE, per-step action masks.
- Episode-level win/loss signal for team reward
- Difficulty scaling via unit type distributions and team size

Repository: [SMACv2-repository](https://github.com/oxwhirl/smacv2)
Paper: Ellis et al., "SMACv2: An Improved Benchmark for Cooperative Multi-Agent Reinforcement Learning", NeurIPS 2023

<img width="2400" height="1288" alt="Image" src="https://github.com/user-attachments/assets/2489840d-a655-4103-8fbe-160b0ec35007" />

### Why This Matters for MOSAIC

1. SMAC v1 is saturated. State-of-the-art algorithms (MAPPO, QMIX) already achieve near-100% win rates on most SMAC v1 maps. The research community has largely moved to SMACv2 as the standard evaluation. Adding v1 without v2 would make MOSAIC comparisons immediately outdated.
2. Procedural generation tests real generalisation. Because unit positions and types are randomised each episode, a policy that scores well on SMACv2 must have actually learned cooperative behaviour rather than memorised a fixed sequence of moves. This is exactly the kind of rigorous evaluation that MOSAIC's cross-paradigm mixing experiments require.
3. Same CTDE structure as SMAC v1. SMACv2 exposes the same per-agent local observation + global state tensor split. MOSAIC's Operator system can route these identically: local obs to decentralised actors, global state to a centralised critic. No architectural changes needed relative to a SMAC v1 integration.
4. Heterogeneous unit mixing maps onto cross-paradigm vision. With randomised unit type compositions, a natural MOSAIC experiment is: assign RL agents to one unit type, an LLM agent to another, and measure whether cross-paradigm teams can adapt to novel compositions episode-to-episode — a test no fixed-map benchmark can provide.
5. Action masking is required and surfaced. SMACv2 exposes get_avail_agent_actions() per step, exactly as SMAC v1 does. MOSAIC needs to forward these masks through the Operator interface to prevent illegal actions and for correct credit assignment in CTDE algorithms.

**Comment by Abdulhamid97Mousa** [comment]
### SMACv2 is already integrated in MOSAIC alongside SMAC v1, and we shipped both together precisely for the reason you identified. 

https://github.com/user-attachments/assets/dc83a7d3-8088-49f5-a1ad-05acece24b50

### What was implemented:
- SMACv2Adapter sharing a base class with SMACAdapter, covering all three procedural map families: protoss, terran, zerg with configurable team sizes (5v5, 10v10, 20v20).
- Unit start positions and type compositions randomised every episode.  the adapter resets cleanly each episode with a new procedural composition, no state leakage.
- Per-step action masks forwarded through the Operator interface via get_avail_agent_actions(), identical to v1.
- Global state tensor exposed for CTDE routing.  local obs to decentralized actors, global state to a centralized critic.
- Episode-level win/loss signal propagated to the training worker

**Comment by Abdulhamid97Mousa** [comment]
### MAPPO training on SMACv2 is also fully working. Here is a live session on terran_20_vs_23 running inside MOSAIC

https://github.com/user-attachments/assets/a752f29d-6b40-4a36-adea-21f3bc09069c

### Hyperparameters used (reproducible):

| Parameter           | Value      |
|---------------------|-----------------|
| method              | MAPPO    |
| env                 | smacv2         |
| env_id              | terran_20_vs_23 |
| dl_toolbox          | torch           |
| device              | cuda:0          |
| parallels           | 4               |
| running_steps       | 1_000_000       |
| seed                | 1               |
| env_seed            | 1               |
| smac_render_mode    | 3d   |
| smac_render_size    | 1024   |
| fastlane_enabled    | True            |
| fastlane_video_mode | grid            |
| fastlane_grid_limit | 4               |
| track_tensorboard   | True            |

To reproduce, launch xuance_worker from the MOSAIC GUI with the above settings, or pass them directly via the XuanCe YAML config. Any of the 15 SMACv2 maps (protoss_*, terran_*, zerg_*) can be substituted for env_id

**Note: please select the 3D GPU accelerated rendering not heat map if you want rich visuals**
